Adding patch from cdk to fix snapd_refresh smash by subordinate charm

hyperbolic2346 opened the pull request:
Also had to add a check for space only. I wonder if that needs to be ported to cdk. I'll check in the morning...

https://github.com/juju-solutions/kubernetes/pull/212

hyperbolic2346 committed:
Adding patch from cdk to fix snapd_refresh smash by subordinate charm

Also had to add a check for space only. I wonder if that needs to be ported to cdk. I'll check in the morning...

hyperbolic2346 committed:
Fixing regular expression flake8 issue

kwmonroe commented on reactive/etcd.py:
```diff
@@ -428,6 +428,12 @@ def process_snapd_timer():
     # Get the current snapd refresh timer; we know layer-snap has set this
     # when the 'snap.refresh.set' flag is present.
     timer = snap.get(snapname='core', key='refresh.timer').decode('utf-8')
+    if not timer or timer.isspace():
```
`isspace` works here because the timer gets set to an empty string by the subordinate, which layer-snap eventually returns as `\n`:

```
sudo snap set core refresh.timer=''
...
>>> subprocess.check_output(['sudo', 'snap', 'get', 'core', 'refresh.timer']).decode('utf-8')
'\n'
```

How about we check `timer` against the configured value, as in:
```
if not timer or timer != hookenv.config('snapd_refresh')
```

This would mean that etcd would be in charge of the host snap timer, effectively ignoring any subordinate refresh config, but maybe that's ok?

kwmonroe replied:
On second thought, I think it's more important to set the refresh timer to whatever it was before the subordinate smashed it.  Blindly falling back to the charm config may not be what we want.

For example, deploy etcd today, Thursday the 18th.  By default, etcd configures a 'max' refresh timer.  This is translated by layer-snap to "wed3" to signify that the snap should not check for refreshes until the 3rd Wednesday (effectively  a month from yesterday).

If we add a subordinate with no refresh configured to the etcd unit in a couple weeks, it will nullify the core snap refresh timer, and the leader will fall into this method.  With the current patch in place, etcd will set a new timer based on config (still 'max'), which layer-snap will translate into something other than the previously set 'wed3' timer.

All that to say, let's keep checking timer for None or isspace here. If that condition is true, check for previously set timers with `leader_get` and only fall back to config if we don't have a previous value.

hyperbolic2346 replied:
Comparing against the config value will catch a situation where the subordinate has a value and the charm does as well. Who should win there? I would vote the main charm.

kwmonroe replied:
Yeah, I agree that the principal charm should always win.

However, since our config defaults to 'max', it's not as simple as comparing the current timer to config since the current timer will never say 'max'.

kwmonroe committed:
timer tweaks to consider leader-set and config when resetting

kwmonroe committed:
add reasonable flake8 line length

kwmonroe committed:
make empty configured refresh timer valid on reset

When resetting the snap refresh timer (because a subordinate clobbered it),
make sure we allow for an empty string as the configured timer. Some
operators may want to disable the refresh timer, which is done by setting
the config to ''.

kwmonroe committed:
belt/suspenders make sure a valid timer string is set

hyperbolic2346 reviewed (commented):
What if timer is not "", but it isn't the value we want? Like they deploy the subordinate with a value for refresh that clashes with the one we want?

kwmonroe committed:
readme updates

- format tweaks
- add --wait to action commands
- add snapd_refresh limitation, noting best practice

kwmonroe committed:
update config description with truth

hyperbolic2346 commented on README.md:
```diff
@@ -349,16 +360,15 @@ a single node due to the nature of how replicas work in etcd.
-- Charles Butler &lt;[charles.butler@canonical.com](mailto:charles.butler@canonical.com)&gt;
-- Mathew Bruzek  &lt;[mathew.bruzek@canonical.com](mailto:mathew.bruzek@canonical.com)&gt;
+The `etcd` charm is free and open source software created by the
+~containers team at Canonical.
```
No link or mailing list or something? launchpad group?

kwmonroe committed:
update contact info in the readme
